Fix kInvalidFileID not matching file_id_t

floam opened the pull request:
In testing I noticed at runtime what appeared to be `time_t` timestamps being converted  to `uint32_t` for `kInvalidFileId`.

45dfa2d8640ab94c86ba97d5dc15db87badc67fb added `mod_seconds` and `mod_nanosecondas` to file_id_t:

```c++
 struct file_id_t
 {
      dev_t device;
      ino_t inode;
      uint64_t size;
      time_t change_seconds;
      long change_nanoseconds;
 +    time_t mod_seconds;
 +    long mod_nanoseconds;
      uint32_t generation;
]
```

But this wasn't updated:

```c++
const file_id_t kInvalidFileID = {(dev_t)-1LL, (ino_t)-1LL, (uint64_t)-1LL, -1, -1, (uint32_t)-1};
```

Where it has `(uint32_t)-1` it still assumes that is the `generation` field but that is `mod_seconds` now!

It's basically used to initialize file_id_t prior to being overwritten by `stat()` values (if all goes well).

floam commented:
```
#if defined(__APPLE__) || defined(__DragonFly__) || defined(__FreeBSD__) ||  defined(__OpenBSD__) || defined(__NetBSD__)
    result.generation = buf->st_gen;
#else
    result.generation = 0;
#endif
```

There must be a good reason to avoid st_gen on Linux. But I don't understand what it is. Wouldn't this help avoid the collisions 45dfa2d8640ab94c86ba97d5dc15db87badc67fb is trying to address?

I may not understand filesystems.

floam commented:
Given my confusion above and this: 

> The solution is to manually set the nanosecond field of the mtime timestamp to
> something unlikely to be duplicated, like a random number, or better yet, the
> current time (with nanosecond precision). This is more in the spirit of the
> timestamp, and it means we're around a million times less likely to collide.
> This seems to fix the tests.

... it could be this patch is boneheaded and it was exactly the intended behavior that I ran into. @ridiculousfish ?

floam commented:
Deleted some comments about st_gen - I was mistaken in thinking it worked on Linux.

krader1961 commented:
Not only is `st_gen` not available on Linux it's not usable on BSD systems except by root. From `man 2 stat`:

```
The file generation number, st_gen, is only available to the super-user.
```

krader1961 commented:
@ridiculousfish: What led you to believe the `generation` member of `struct stat` was usable on BSD based systems like Mac OS X? According to the man page it's only exposed to the super-user. I added this diagnostic printf and it reports zero on my OS X server:

```
fprintf(stderr, "WTF buf->st_gen == %d\n", buf->st_gen);
```

krader1961 commented on src/wutil.cpp:
```diff
@@ -564,7 +564,7 @@ file_id_t file_id_t::file_id_from_stat(const struct stat *buf)
     result.generation = buf->st_gen;
 #else
-    result.generation = 0;
+    result.generation = NULL;
```
This is wrong because the generation member of the struct is a uint32_t not a pointer. But as I've hinted in my other comments regarding this change I think we should just drop the generation member.

floam replied:
On Linux there is may be no such thing as an inode generation.

NULL in C++ is explicitly defined as 0, so I considered it an aesthetic difference. I'm certainly not an old hand here so I defer to your wisdom. I put it there to be explicit with myself that buf->st_gen does not exist on Linux and that member is no good. 

Makes no difference to me if it's set to zero by NULL or 0 in the end.

krader1961 replied:
`NULL` should only be used with respect to pointers. The whole point of `NULL` was that, at least in the past, there were computer architectures that required a non-zero value to represent a pointer that could not be dereferenced. You'll find a lot of B.S. regarding this if you google questions like "c++ null pointer value". That's because people who haven't been programming for 30+ years and are therefore unlikely to have encountered unusual architectures like the 24 bit word length computers sold by Harris Computer Systems in the 1980's are unaware of the relevant history.

floam replied:
I don't care to pay to access the actual standards as written, I use basically whatever Dash on OS X has on hand for most languages. For C++ I have [this resource pop up if I type in NULL](http://en.cppreference.com/w/cpp/types/NULL) which aggregates information on how the standard has evolved over time.

krader1961 replied:
The document you referred to begins with

> The macro NULL is an implementation-defined null pointer constant

Notice the word "pointer" in that sentence. It is not idiomatic to use NULL in a non-pointer context regardless of the fact that in practice, today, it will always evaluate to zero.

floam replied:
The place to look if you really want to see something besides zeros there will be NFS and userspace filesystems.

The standard says that field will be there and it's accessed by standard means so I see no reason not to store the data from the struct even if it might be zero'd.

floam commented:
> What led you to believe the generation member of struct stat was usable on BSD based systems like Mac OS X? 

What the manual page says and the truth can be different things. Play with your test code.

krader1961 commented:
> What the manual page says and the truth can be different things. Play with your test code.
> And you're only talking about OS X. I said BSD.

Which is why I built a fish with a debug printf. And it shows that the generation is zero for non-root users. And OS X is BSD based. There might be other BSD variants that populate that value for non-root users but the code as written assumes it is available on OS X.

Also, since you're making a non-trivial, non-janitorial, change I'd be happier if it was tied to an open issue where the problem is described and the rationale for the change is explained.

floam commented:
This problem this PR addresses is that kInvalidFileID is missing two members, causing a timestamp to lose half it's precision when cast to uint32_t, (and the last two members are left to chance should stat() fail, maybe not an issue, I haven't figured that out.)

krader1961 commented:
> This problem this PR addresses is that kInvalidFileID is missing two members....

I really do appreciate that you took the time to write a pull-request comment to that effect so people reviewing your PR would have context for the change. My point is that it would be better for people in the future who are trying to understand why this change was made if that comment was in an issue and the git commit comment had a "Fixes #1234" line referencing that issue.

Please don't think I'm singling you out regarding this practice. Long-time contributors like faho, zanchey, and ridiculous fish have also been doing this. And I'm going to yell at them if they continue to do it :-)

floam commented:
I've removed the line not passing kraderlint :) 

I've updated the commit message to express the change more succinctly and I referenced the PR # here.

Are you saying at this point we should go make a GitHub issue for this?

ridiculousfish commented:
> What led you to believe the generation member of struct stat was usable on BSD based systems like Mac OS X?

Well they kind of bury that fact :( Anyways we can remove it.

floam commented:
Sorry for that mess, there is a huge mess (for me here, at least) of commits above this - I thought they were commits in this PR for a while and I kept force pushing to fix it and doing different stuff to squash them away and change HEAD -- actually it's a bunch of "referenced this pull request" notifications.

Just made it worse.

The actual commit of the branch is clean. Whoops. Don't know how to wipe those!
